rust/src/main.rs

```diff
@@ -45,15 +46,47 @@ fn get_logging_directory() -> Result<PathBuf, io::Error> {
     }
 }
 
-fn path_for_log_file<P: AsRef<Path>>(filename: P) -> Result<PathBuf, io::Error> {
-    let mut logging_directory = get_logging_directory()?;
-    // Create the logging directory
-    fs::create_dir_all(&logging_directory)?;
-    logging_directory.push(filename.as_ref());
-    Ok(logging_directory)
+/// This function tries to create the parent directories for a file
+///
+/// It wraps around the `parent()` function of `Path` which returns an `Option<&Path>` and
+/// `fs::create_dir_all` which returns an `io::Result<()`.
+///
+/// This allows you to use `?`/`try!()` to create the dir and you recive the additional custom error for when `parent()`
+/// returns nothing.
+///
+/// # Errors
+/// This can return an `io::Error` if `fs::create_dir_all` fails or if `parent()` returns `None`.
+/// See `Path`'s `parent()` function for more details.
+/// # Examples
+/// ```
+/// use std::path::PathBuf;
+/// let path_with_file = PathBuf::from("/some/directory/then/file");
+/// assert_eq!(Some(OsStr::new("file")), path_with_file.file_name());
+/// assert_eq!(create_log_directory(path_with_file).is_ok(), true);
+///
+/// let path_with_other_file = PathBuf::from("/other_file");
+/// assert_eq!(Some(OsStr::new("other_file")), path_with_other_file.file_name());
+/// assert_eq!(create_log_directory(path_with_file).is_ok(), true);
+///
+/// // Path that is just the root or prefix:
+/// let path_without_file = PathBuf::from("/");
+/// assert_eq!(None, path_with_file.file_name());
+/// assert_eq!(create_log_directory(path_with_file).is_ok(), false);
+/// ```
+fn create_log_directory(path_with_file: &PathBuf) -> Result<(), io::Error> {
+    let log_dir = path_with_file.parent().ok_or(io::Error::new(
+        io::ErrorKind::InvalidInput,
+        format!(
+            "Unable to get the parent of the following Path: {}, Your path should contain a file name",
+            path_with_file.display(),
+        ),
+    ))?;
+    // Try to create the directory.
```

Comment:
At the risk of being too nitty, an observation I have in a few places in this PR is that it might be slightly _over_ commented. I appreciate that you're trying to write 'good code', and that's an excellent thing to aspire to, but it's easy to get a little overzealous sometimes.

I think comments are important, but the one case where I like to discourage them is when reading the comment is actually _slower_ than reading the code. In this instance, the name of the function tells us all we need to know, and the `?` tells us this is failable.

In any case, I would normally not mention this but you've been pretty receptive to the review process so far, and I think this advice might be helpful. This is not a blocking problem by any means, I'm not asking for a change, just explaining what I see. 😎

Reply:
I usually under comment so I've probably put a few too many here and there.
Happy to remove them as I agree in these cases the code is already self explanatory.

Keep up the nitpicking!